pandas_sphinx_theme/__init__.py

```diff
@@ -15,52 +15,97 @@
-
-
-def convert_docutils_node(list_item, only_pages=False):
+# Sphinx monkeypatch for adding toctree objects into context.
+# This converts the docutils nodes into a nested dictionary that Jinja can
+# use in our templating.
+
+
+def docutils_node_to_jinja(list_item, only_pages=False):
+    """Convert a docutils node to a structure that can be read by Jinja.
+
+    Parameters
+    ----------
+    list_item : docutils list_item node
+        A parent item, potentially with children, corresponding to the level
+        of a TocTree.
+    only_pages : bool
+        Only include items for full pages in the output dictionary. Exclude
+        anchor links (TOC items with a URL that starts with #)
+
+    Returns
+    -------
+    nav : dict
+        The TocTree, converted into a dictionary with key/values that work
+        within Jinja.
+    """
     if not list_item.children:
         return None
+
+    # We assume this structure of a list item:
+    # <list_item>
+    #     <compact_paragraph >
+    #         <reference> <-- the thing we want
     reference = list_item.children[0].children[0]
     title = reference.astext()
     url = reference.attributes["refuri"]
     active = "current" in list_item.attributes["classes"]
 
+    # If we've got an anchor link, skip it if we wish
     if only_pages and '#' in url:
         return None
 
+    # Converting the docutils attributes into jinja-friendly objects
     nav = {}
     nav["title"] = title
     nav["url"] = url
-    nav["children"] = []
     nav["active"] = active
 
+    # Recursively convert children as well
+    # If there are sub-pages for this list_item, there should be two children:
+    # a paragraph, and a bullet_list.
+    nav["children"] = []
     if len(list_item.children) > 1:
-        for child_item in list_item.children[1].children:
-            child_nav = convert_docutils_node(child_item, only_pages=only_pages)
+        # The `.children` of the bullet_list has the nodes of the sub-pages.
+        subpage_list = list_item.children[1].children
+        for sub_page in subpage_list:
+            child_nav = docutils_node_to_jinja(sub_page, only_pages=only_pages)
             if child_nav is not None:
                 nav["children"].append(child_nav)
-
     return nav
 
 
 def update_page_context(self, pagename, templatename, ctx, event_arg):
     from sphinx.environment.adapters.toctree import TocTree
 
-    def get_nav_object(**kwds):
-        """Return a list of nav links that can be accessed from Jinja."""
+    def get_nav_object(maxdepth=None, **kwargs):
+        """Return a list of nav links that can be accessed from Jinja.
+
+        Parameters
+        ----------
+        maxdepth: int
+            How many layers of TocTree will be returned
+        kwargs: key/val pairs
+            Passed to the `TocTree.get_toctree_for` Sphinx method
+        """
+        # The TocTree will contain the full site TocTree including sub-pages.
```

Comment:
Given the collapse=True below, it's actually not the full site TocTree I think? (but only full up to the level of the requested page, and then only the further subpages (depending on maxdepth) on the requested page)

Reply:
We might want to make this configurable (I mean, from the template) by moving the collapse=True into the signature of this function, so when calling it from the template you can override the default

Reply:
I think it's a good idea. Especially if we want to make the UI itself more configurable it'd be useful to think about at least these three use-cases when designing these functions:

1. The current theme setup (top bar w/ first-level headers)
2. A setup where the left sidebar has *all* levels in it (perhaps without the topbar)
3. A setup where the left sidebar has all sub-levels shown instead of only showing "active" ones, or with a collapsible arrow next to non-active ones.

What do you think? And if you're +1 on that, I think we should open up a new issue to discuss improving these functions, beyond documenting their functionality as this PR does.

Reply:
Yes, I would like to see support for all those items you mentioned.